python.pipInstallHook: avoid producing wrong direct_url.json file

yajo opened the pull request:
Fix problem detected in https://github.com/NixOS/nixpkgs/pull/229472#issuecomment-1544410416.

Since bff6c67911ed74cf560bd2b00fb1da32bd2fed63, now pname is expected to match the name of the package that is built by Python standard tools.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

yajo committed:
python.pipInstallHook: avoid producing wrong direct_url.json file

When installing many python packages, a `direct_url.json` file appeared in the lib directory. Example:

```sh
➤ nix build nixpkgs/29755fec55e58a315b517d431b2be261772f2b80#python3Packages.flask

➤ cat result/lib/python3.10/site-packages/Flask-2.2.3.dist-info/direct_url.json
{"archive_info": {}, "url": "file:///build/Flask-2.2.3/dist/Flask-2.2.3-py3-none-any.whl"}⏎
```

As you can see, that file contains a wrong reference to `/build`.

In https://discuss.python.org/t/pep-610-usage-guidelines-for-linux-distributions/4012/4 there's an explanation on how to avoid this. Here, I'm implementing that change for nixpkgs.

@moduon MT-1075

yajo committed:
python310Packages.lz4: fix build

Fix problem detected in https://github.com/NixOS/nixpkgs/pull/229472#issuecomment-1544410416.

Since bff6c67911ed74cf560bd2b00fb1da32bd2fed63, now pname is expected to match the name of the package that is built by Python standard tools.

FRidh commented:
hmm, yes, I am not sure whether we can really rely on pname. We need to test this and merge after branch-off. If everything is working well we can always choose to backport to stable.

vcunat commented:
This was discussed in https://matrix.to/#/#staging:nixos.org even before the revert happened, but it doesn't solve the whole problem and it seems too late for such changes in 23.05.

vcunat commented:
`staging` at this point won't reach 23.05 anymore, so it could be used already.

vcunat commented:
For reference, multiple python packages from this list look like suffering from this, at a glance:
https://malob.github.io/nix-review-tools-reports/nixpkgs:staging-next/nixpkgs_staging-next_1795115.html

yajo commented:
Something weird to me is that https://github.com/NixOS/nixpkgs/pull/229472 actually got green ✅ CI. There's probably something specific to nixpkgs that I'm misssing, but how am I supposed to know that a PR breaks things before merging (without toasting my laptop)? 🤔

vcunat commented:
These kinds of changes cause rebuild of most packages.  That's very expensive; way too much for just a CI.  I could set up a separate jobset on hydra.nixos.org for this, so that it can get somewhat stabilized before merging.  That is, assuming there's consensus that this is the way to go.

vcunat commented:
Though maybe it's an overkill, if it's only expected to be such trivial renames and mostly for packages that fail themselves... in that case I'd expect it could be handled during `staging-next` stabilization (but not for 23.05 anyway).

yajo commented:
Yes, that's basically the fix. You can see the details in https://github.com/NixOS/nixpkgs/pull/229472#issue-1692348730, but basically, instead of installing the whl file directly, we pass the directory name and the package name, and let python find the correct wheel. That's the official way of getting rid of the buggy `direct_url.json`. Thus, the `pname` must match the python package name.

I understand the need to be stable at this point, and I don't mind if this doesn't enter 23.05... but I'd just like to ask: How should I proceed now with this fix?

Thanks!

yajo commented:
Now that 23.05 has born, I've rebased and cherry-picked https://github.com/NixOS/nixpkgs/commit/bff6c67911ed74cf560bd2b00fb1da32bd2fed63 again. This should be mergeable now.

If there's any way to know if this patch is breaking other packages, please tell me. Otherwise, this one should be ready to merge IIUC. Thanks!

yajo commented:
Hi there!

This patch was approved some time ago and merged in https://github.com/NixOS/nixpkgs/pull/229472. Please read that description if you haven't, to understand how this is important. 

Then it was reverted in https://github.com/NixOS/nixpkgs/pull/229472#issuecomment-1551131933 due to the release rush.

I think there should be a review here now that there's plenty of time until the next release, so we can find and fix any possibly broken builds in the mean time.

Would somebody please do the review? Thanks!

SuperSandro2000 reviewed (changes requested):
I think we might now have naming conflicts if the internal package name is uppercase like for Twisted, see https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md?plain=1#L1940-L1945

```
 ➜ nix build .#python3.pkgs.twisted.dist

 ➜ ls -lah result-dist
lrwxrwxrwx sandro users 75 B Tue Jul 18 18:42:11 2023  result-dist ⇒ /nix/store/bjc37hqna1g8374d6hp9aq4069dinva9-python3.10-twisted-22.10.0-dist
```

Also we should really add some helpful error messages otherwise people might be really confused if such a case happens for another package.

FRidh commented:
Yes, I actually think this is not a good solution. As far as I recall there is no guarantee whatsoever that a package name on PyPI will match with a wheel name.

yajo committed:
python: extend audit tmpdir for direct_url.json files

This makes sure the problem we're fixing gets fixed.

yajo committed:
python.pipInstallHook: support wheelName

As seen in https://github.com/NixOS/nixpkgs/pull/232414#pullrequestreview-1535468883, some packages get built to a different name than nixpkgs' standard pname.

Those packages can now use `wheelPname` to get built.

Added to Twisted, just to test it out.

yajo commented:
I pushed some changes, please review again.

Now I have extended the audit tmpdir function. I always wondered why this wasn't detected here: https://github.com/NixOS/nixpkgs/blob/3fa49138c9e24841bc68f73e9fb80998d886abab/pkgs/build-support/setup-hooks/audit-tmpdir.sh#L16

That test would break almost all python packages if it weren't because I push the fix here too. 😆 But you can run it if you want, to see how `/build` was slipping into the built packages.

Maybe it's not so problematic after all, I'm not sure. According to https://packaging.python.org/en/latest/specifications/direct-url/, it's just metadata.  🤷🏼‍♂️ But in general I think nothing from /build should land in the drv output.

I fixed a couple of derivations, including Twisted, making use of the new `wheelPname` attribute. Also I added a helpful error message.

SuperSandro2000 commented on pkgs/development/python-modules/twisted/default.nix:
```diff
@@ -48,11 +48,12 @@ buildPythonPackage rec {
 
   src = fetchPypi {
-    pname = "Twisted";
+    pname = wheelPname;
```
```suggestion
    pname = wheelName;
```
not sure if that is better or if we should rename the attr in fetchPypi

yajo replied:
I think `wheelPname` is better because it is more intuitive that you need to put the value that you would be putting in `pname` and not in `name`. The wheel files that are generated include the pname and version (aka name), so here we really care about the pname part.

yajo replied:
> not sure if that is better or if we should rename the attr in fetchPypi

Most times the value will be the same. I can add that attribute there, and assert you're setting only one of both. Since those derivations are FODs, changes there shouldn't produce mass rebuilds.

SuperSandro2000 commented on pkgs/development/interpreters/python/hooks/pip-install-hook.sh:
```diff
@@ -22,3 +25,20 @@ if [ -z "${dontUsePipInstall-}" ] && [ -z "${installPhase-}" ]; then
 fi
+
+
+pipAuditTmpdir() {
```
There is also a tool called pip-audit which makes this a bit confusing

yajo replied:
I was trying to follow the terminology from here because it's mostly extending that safety net: https://github.com/NixOS/nixpkgs/blob/3fa49138c9e24841bc68f73e9fb80998d886abab/pkgs/build-support/setup-hooks/audit-tmpdir.sh#L12

But just tell me what name you prefer, I can change it, no problem.

FRidh commented:
We should take a step back here and look at what the problem is that we're trying to solve.

[PEP 610](https://peps.python.org/pep-0610/) describes how frontends should create a `direct_url.json` file when installing wheels. This file is causing trouble for us, because it can contain references we don't like. These references aren't really used, they're for provenance. We care about provenance, but provenance is already part of our Nix expressions. So, for as far I can tell, we may as well delete these files.

The *only* case we're interested in them, is for [PEP 662 editable installations](https://peps.python.org/pep-0662/). But that's only when using the phase specific for editable installations. For regular installations, we can still delete the file.

Thus, I suggest we always remove the file from the `dist-info` folder.

yajo commented:
It's another possible fix. It just felt more safe for me to let do pip its stuff without interfering.

yajo commented:
Does this still make sense after https://github.com/NixOS/nixpkgs/pull/248866 was merged? 🤔
